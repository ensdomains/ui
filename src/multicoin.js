import bs58 from 'bs58'

export function encodeToBytes(input, encoding) {
  switch (encoding) {
    case 'hex-checksum':
      return input
    case 'base58check':
      const bytes = bs58.decode(input)
      return bytes
    default:
      throw `Unsupported base ${base}`
  }
}

export function decodeFromBytes(bytes, encoding) {
  switch (encoding) {
    case 'hex-checksum':
      return '0x' + bytes.toString('hex')
    case 'base58check':
      return bs58.encode(bytes)
    default:
      throw `Unsupported encoding: ${encoding}`
  }
}
