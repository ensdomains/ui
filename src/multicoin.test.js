import { encodeToBytes, decodeFromBytes } from './multicoin'

describe('encodeToBytes', () => {
  it('encodes bitcoin hash to correct hex', () => {
    const hash = '1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa'
    const result = encodeToBytes(hash, 'base58check')
    expect(result).toMatchSnapshot()
  })

  it('encodes litecoin has to correct hex', () => {
    const hash = 'LeJ4KvE7QhURHecZfNH7Kj6CDRwgM6rEGN'
    expect(encodeToBytes(hash, 'base58check')).toMatchSnapshot()
  })
})

describe('decodeFromBytes', () => {
  it('decodes bitcoin hex to correct base58 string', () => {
    const address = '1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa'
    expect(
      decodeFromBytes(
        Buffer.from([
          0,
          98,
          233,
          7,
          177,
          92,
          191,
          39,
          213,
          66,
          83,
          153,
          235,
          246,
          240,
          251,
          80,
          235,
          184,
          143,
          24,
          194,
          155,
          125,
          147
        ]),
        'base58check'
      )
    ).toEqual(address)
  })

  it('decodes litecoin hex to correct base58 string', () => {
    const address = 'LeJ4KvE7QhURHecZfNH7Kj6CDRwgM6rEGN'
    expect(
      decodeFromBytes(
        Buffer.from([
          48,
          209,
          49,
          29,
          153,
          29,
          5,
          220,
          128,
          29,
          99,
          172,
          195,
          193,
          66,
          50,
          16,
          18,
          239,
          171,
          237,
          55,
          130,
          240,
          135
        ]),
        'base58check'
      )
    ).toEqual(address)
  })
})

describe('encodes and decodes back and forth', () => {
  it('encodes/decodes bitcoin hash', () => {
    const hash = '1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa'
    const result = encodeToBytes(hash, 'base58check')
    expect(decodeFromBytes(result, 'base58check')).toEqual(hash)
  })
  it('encodes/decodes litecoin hash', () => {
    const hash = 'LeJ4KvE7QhURHecZfNH7Kj6CDRwgM6rEGN'
    const result = encodeToBytes(hash, 'base58check')
    expect(decodeFromBytes(result, 'base58check')).toEqual(hash)
  })
})
