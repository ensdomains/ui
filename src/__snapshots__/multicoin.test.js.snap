// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decodeFromBytes decodes litecoin hex to correct base58 string 1`] = `"1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa"`;

exports[`encodeToBytes encodes bitcoin hash to correct hex 1`] = `
Object {
  "data": Array [
    0,
    98,
    233,
    7,
    177,
    92,
    191,
    39,
    213,
    66,
    83,
    153,
    235,
    246,
    240,
    251,
    80,
    235,
    184,
    143,
    24,
    194,
    155,
    125,
    147,
  ],
  "type": "Buffer",
}
`;

exports[`encodeToBytes encodes litecoin has to correct hex 1`] = `
Object {
  "data": Array [
    48,
    209,
    49,
    29,
    153,
    29,
    5,
    220,
    128,
    29,
    99,
    172,
    195,
    193,
    66,
    50,
    16,
    18,
    239,
    171,
    237,
    55,
    130,
    240,
    135,
  ],
  "type": "Buffer",
}
`;

exports[`encodeToHex encodes bitcoin hash to correct hex 1`] = `
Object {
  "data": Array [
    0,
    98,
    233,
    7,
    177,
    92,
    191,
    39,
    213,
    66,
    83,
    153,
    235,
    246,
    240,
    251,
    80,
    235,
    184,
    143,
    24,
    194,
    155,
    125,
    147,
  ],
  "type": "Buffer",
}
`;
